### PR TITLE
Python path queries

### PR DESCRIPTION
**Pull request by markshannon**

### CODE REVIEW

File: python/ql/src/Security/CWE-798/HardcodedCredentials.ql [hunk 1]
```diff
@@ -3,7 +3,7 @@
  * @description Credentials are hard coded in the source code of the application.
  * @kind problem
  * @problem.severity error
- * @precision medium
+ * @precision high
```
Comment (felicity-semmle) [inline python/ql/src/Security/CWE-798/HardcodedCredentials.ql]: This is going to change the behaviour of this query on LGTM. Please add a change note.

File: python/ql/src/Security/CWE-022/PathInjection.ql [hunk 2]
```diff
@@ -1,7 +1,7 @@
 /**
  * @name Uncontrolled data used in path expression
  * @description Accessing paths influenced by users can allow an attacker to access unexpected resources.
- * @kind problem
+ * @kind path-problem
```
Comment (felicity-semmle) [inline python/ql/src/Security/CWE-022/PathInjection.ql]: We should probably have a change note about the conversion of some queries to path queries.

File: python/ql/test/query-tests/Security/CWE-078/CommandInjection.expected [hunk 3]
```diff
@@ -1,3 +1,19 @@
...
+| command_injection.py:19:22:19:34 | shell command | command_injection.py:17:13:17:24 | Taint {externally controlled string} at command_injection.py:17 | command_injection.py:19:22:19:34 | Taint [externally controlled string] at command_injection.py:19 | This command depends on $@. | command_injection.py:17:13:17:24 | flask.request.args | a user-provided value |
+| command_injection.py:25:22:25:36 | OS command first argument | command_injection.py:24:11:24:22 | Taint {externally controlled string} at command_injection.py:24 | command_injection.py:25:22:25:36 | Taint [externally controlled string] at command_injection.py:25 | This command depends on $@. | command_injection.py:24:11:24:22 | flask.request.args | a user-provided value |
+| command_injection.py:25:22:25:36 | OS command first argument | command_injection.py:24:11:24:22 | Taint {externally controlled string} at command_injection.py:24 | command_injection.py:25:22:25:36 | Taint sequence[externally controlled string][0] at command_injection.py:25 | This command depends on $@. | command_injection.py:24:11:24:22 | flask.request.args | a user-provided value |
```
Comment (taus-semmle) [inline python/ql/test/query-tests/Security/CWE-078/CommandInjection.expected]: I wonder why this final result was duplicated.

Comment (taus-semmle) [reply python/ql/test/query-tests/Security/CWE-078/CommandInjection.expected]: Ah, I see the difference. One has `[0]` appended to the taint.

File: python/ql/src/Security/CWE-022/PathInjection.ql [hunk 4]
```diff
@@ -25,7 +26,8 @@ import semmle.python.web.HttpRequest
 import semmle.python.security.injection.Path
 
 
-from TaintSource src, TaintSink sink
-where src.flowsToSink(sink)
 
-select sink, "This path depends on $@.", src, "a user-provided value"
+from TaintedNode srcnode, TaintedNode sinknode, TaintSource src, TaintSink sink
+where src.flowsToSink(sink) and srcnode.getNode() = src and sinknode.getNode() = sink
```
Comment (taus-semmle) [inline python/ql/src/Security/CWE-022/PathInjection.ql]: I feel like the `srcnode.getNode() = src and sinknode.getNode() = sink` construction is a bit clunky. Could we make this into a method (`.getSource()`) on `TaintedNodes`, etc. and then simply use `srcnode.getSource()` in the `select` statement?

Comment (markshannon) [reply python/ql/src/Security/CWE-022/PathInjection.ql]: That makes sense. I've changed the general form from
```
from TaintedNode srcnode, TaintedNode sinknode, TaintSource src, TaintSink sink
where src.flowsToSink(sink) and srcnode.getNode() = src and sinknode.getNode() = sink
select sink, srcnode, sinknode, ..., src, ...
```
to 
```
from TaintedPathSource src, TaintedPathSink sink
where src.flowsTo(sink)
select sink.getSink(), src, sink,..., src.getSource(), ...
```

Comment (taus-semmle) [reply python/ql/src/Security/CWE-022/PathInjection.ql]: Perfect!

File: python/ql/src/semmle/python/security/Paths.qll [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+private TaintedNode next_sibling(TaintedNode child) {
+    child.getASuccessor() = result and
+    child.getContext() = result.getContext()
+}
+
+query predicate parents(TaintedNode child, TaintedNode parent) {
+    child = first_child(parent) or
+    exists(TaintedNode prev |
+        parents(prev, parent) and
+        child = next_sibling(child)
```
Comment (aschackmull) [inline python/ql/src/semmle/python/security/Paths.qll]: Post-merge drive-by comment: This looks wrong.
